File: scripts/projections/featured.py
"""data.featured.json projection — per DATA_ARCHITECTURE.md §6.8.

Status: Future (skipped by default)
Consumer: mobile homepage hero ("today's pick" / "high AI risk occupations")
Shape: { generated_at, strategy, occupations: [{...full detail...}] }
Size target: < 10 KB gzipped

Strategy is configurable; default 'top_ai_risk_with_workforce' picks 12 records:
    - ai_risk >= 7 AND workers >= 50_000
    - sorted by ai_risk desc, then workers desc

Each entry inlines the full detail (avoid second fetch on mobile first paint).
"""
from __future__ import annotations

import datetime as dt
import json
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from scripts.lib.indexes import Indexes


STRATEGY = "top_ai_risk_with_workforce"
PICK_COUNT = 12


def _candidate(indexes: "Indexes", occ_id: int) -> dict | None:
    occ = indexes.occ_by_id[occ_id]
    score = indexes.latest_score_by_occ.get(occ_id)
    stats = indexes.stats_by_id.get(occ_id)
    if not score or not stats or stats.workers is None:
        return None
    if score.ai_risk < 7 or stats.workers < 50_000:
        return None
    trans = indexes.trans_by_id.get(occ_id)
    return {
        "id": occ_id,
        "title_ja": occ.title_ja,
        "title_en": trans.title_en if trans else None,
        "ai_risk": score.ai_risk,
        "workers": stats.workers,
        "salary_man_yen": stats.salary_man_yen,
        "rationale_ja": score.rationale_ja,
        "summary_ja": occ.description.summary_ja,
        "summary_en": trans.summary_en if trans else None,
        "url": occ.url,
    }


def build(indexes: "Indexes", dist_root: Path) -> dict:
    candidates = [c for c in (_candidate(indexes, oid) for oid in indexes.occ_by_id) if c]
    candidates.sort(key=lambda x: (-x["ai_risk"], -(x["workers"] or 0)))
    picks = candidates[:PICK_COUNT]

    payload = {
        "schema_version": "1.0",
        "generated_at": dt.datetime.now(dt.timezone.utc).isoformat(timespec="seconds"),
        "strategy": STRATEGY,
        "occupations": picks,
    }
    out = dist_root / "data.featured.json"
    out.write_text(json.dumps(payload, ensure_ascii=False, separators=(",", ":")) + "\n", encoding="utf-8")
    return {"file": out, "picked": len(picks), "candidate_pool": len(candidates)}

File: src/data/projections/featured.ts
/**
 * data.featured.json projection — per docs/DATA_ARCHITECTURE.md §6.8.
 *
 * Status: Future (skipped by default — `--enable-future` in Python)
 * Consumer: mobile homepage hero ("today's pick" / "high AI risk occupations")
 *
 * Strategy: 'top_ai_risk_with_workforce' picks 12 records:
 *   - ai_risk >= 7 AND workers >= 50_000
 *   - sorted by ai_risk desc, then workers desc
 *
 * Migrated from scripts/projections/featured.py.
 */
import { writeFile } from 'node:fs/promises';
import { join } from 'node:path';
import type { Indexes } from '../lib/indexes.js';

const STRATEGY = 'top_ai_risk_with_workforce';
const PICK_COUNT = 12;

export interface FeaturedBuildResult {
  files: string[];
  picked: number;
  candidatePool: number;
}

interface CandidateEntry {
  id: number;
  title_ja: string;
  ai_risk: number;
  workers: number;
  salary_man_yen: number | null;
  rationale_ja: string;
  summary_ja: string | null;
  url: string;
}

function nowIso(): string {
  const d = new Date();
  const pad = (n: number) => String(n).padStart(2, '0');
  return (
    `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}` +
    `T${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}+00:00`
  );
}

export async function buildFeatured(
  indexes: Indexes,
  distRoot: string,
): Promise<FeaturedBuildResult> {
  const candidates: CandidateEntry[] = [];
  // Iterate in occById insertion order; matches Python `for oid in indexes.occ_by_id`.
  for (const [occId, occ] of indexes.occById) {
    const score = indexes.latestScoreByOcc.get(occId);
    const stats = indexes.statsById.get(occId);
    if (!score || !stats || stats.workers == null) continue;
    if (score.ai_risk < 7 || stats.workers < 50_000) continue;
    candidates.push({
      id: occId,
      title_ja: occ.title_ja,
      ai_risk: score.ai_risk,
      workers: stats.workers,
      salary_man_yen: stats.salary_man_yen ?? null,
      rationale_ja: score.rationale_ja,
      summary_ja: occ.description.summary_ja ?? null,
      url: occ.url,
    });
  }

  // Sort: ai_risk desc, then workers desc. Stable sort.
  candidates.sort((a, b) => {
    if (b.ai_risk !== a.ai_risk) return b.ai_risk - a.ai_risk;
    return (b.workers ?? 0) - (a.workers ?? 0);
  });
  const picks = candidates.slice(0, PICK_COUNT);

  const payload = {
    schema_version: '1.0',
    generated_at: nowIso(),
    strategy: STRATEGY,
    occupations: picks,
  };

  const outPath = join(distRoot, 'data.featured.json');
  await writeFile(
    outPath,
    JSON.stringify(payload) + '\n',
    'utf-8',
  );

  return {
    files: [outPath],
    picked: picks.length,
    candidatePool: candidates.length,
  };
}
